lib/metric-config-parser/metric_config_parser/experiment.py

```diff
@@ -150,6 +152,24 @@ def bucket_start(self) -> Optional[int]:
 
         return None
 
+    @property
+    def randomization_unit(self) -> Optional[RandomizationUnit]:
+        if hasattr(self.experiment, "bucket_config") and self.experiment.bucket_config is not None:
+            # this will raise a ValueError if the provided randomization_unit is invalid
+            return RandomizationUnit(self.experiment.bucket_config.randomization_unit)
+
+        return None
+
+    @property
+    def analysis_unit(self) -> Optional[AnalysisUnit]:
+        """Retrieve the appropriate analysis unit, which is
+        derived from the experiment's randomization unit.
+        """
+        if self.randomization_unit and self.randomization_unit == RandomizationUnit.GROUP_ID:
```

Comment:
Will this line also raise a ValueError if the randomization_unit is invalid?

Reply:
Yea correct. It'll call the function above to get the randomization unit and so will raise the ValueError.

Reply:
I'll tack on that I thought this was acceptable because the randomization unit really should either be a) undefined (`None`) or b) valid. If it exists and is not a valid value that is bad. One thing I considered was if I should return `None` if the bucket_config is not None but randomization_unit is:
```py
    @property
    def randomization_unit(self) -> Optional[RandomizationUnit]:
        if hasattr(self.experiment, "bucket_config") and self.experiment.bucket_config is not None:
            ########## would add this ##########
            if self.experiment.bucket_config.randomization_unit is None:
                return None
            ########## until here ##########
            # this will raise a ValueError if the provided randomization_unit is invalid
            return RandomizationUnit(self.experiment.bucket_config.randomization_unit)

        return None
```

Reply:
Is the bucket config guaranteed to be provided by the API or are there cases in which it could be missing? I agree with the "if the bucket config is present, require the randomization unit to be valid"  rule

Reply:
Honestly not sure. I'm guessing it will always be there (at least for experiments where we've made it this far and it matters for analysis), but I do tend to get tripped on these things by Normandy stuff sometimes. However, if the bucket_config isn't provided, we fall back to `None` which leads to the default behavior of `AnalysisUnit.CLIENT`. It's really the case where the bucket config exists, but the randomization_unit doesn't that I wasn't sure about. It seems pretty reasonable to consider that invalid, but maybe to confirm I can check in Experimenter next week to see if it's required, and base the decision here on that behavior (i.e., if it's not required by experimenter we could treat `None` as a valid option).

Reply:
Ok confirmed with Jared -- randomization unit should never be None -- the randomization unit depends on the application, and an experiment must always have an application. So I think the current behavior (if bucket_config then randomization unit must not be None) is valid, and throwing an error if that expectation is not met should be fine.

Reply:
Great, thanks for following up on this!